mesontest.py

```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
+
+        jsonlogfile =  open(jsonlogfilename, 'w')
+        logfile = open(logfilename, 'w')
+
+        logfile.write('Log of Meson test suite run on %s.\n\n' %
+                    datetime.datetime.now().isoformat())
+
+        return (logfile, logfilename, jsonlogfile, jsonlogfilename)
+
+    def get_wrapper(self):
+        if self.options.gdb:
+            wrap = ['gdb', '--quiet', '--nh' ]
+            if self.options.repeat > 1:
+                wrap += ['-ex', 'run', '-ex', 'quit']
+            return wrap
+        elif self.options.wrapper:
+            return self.options.wrapper.split()
+        else:
+            return []
+
+    def get_suites(self, tests):
+        return set([test.suite[0] for test in tests])
+
+    def run_tests(self, tests):
         try:
-            if not self.options.verbose:
-                jsonlogfile =  open(jsonlogfilename, 'w')
-                logfile = open(logfilename, 'w')
-                logfile.write('Log of Meson test suite run on %s.\n\n' %
-                            datetime.datetime.now().isoformat())
+            executor = None
+            logfile = None
+            jsonlogfile = None
+            futures = []
+            numlen = len('%d' % len(tests))
+            (logfile, logfilename, jsonlogfile, jsonlogfilename) = self.open_log_files()
+            wrap = self.get_wrapper()
+            single_suite = False
+
+            if len(set([test.suite[0] for test in tests])) == 1:
+                single_suite = True
 
             for i in range(self.options.repeat):
-                for i, test in enumerate(filtered_tests):
-                    if test.suite[0] == '':
+                for i, test in enumerate(tests):
+                    if test.suite[0] == '' or single_suite:
                         visible_name = test.name
                     else:
                         if self.options.suite is not None:
                             visible_name = self.options.suite + ' / ' + test.name
                         else:
                             visible_name = test.suite[0] + ' / ' + test.name
 
-                    if not test.is_parallel:
-                        self.drain_futures(futures)
+                    if self.options.gdb:
+                        test.timeout = None
+                        if len(test.cmd_args):
+                            wrap.append('--args')
+
+                    if not test.is_parallel or self.options.gdb:
+                        self.drain_futures(futures, logfile, jsonlogfile)
                         futures = []
                         res = self.run_single_test(wrap, test)
-                        if not self.options.verbose:
-                            self.print_stats(numlen, filtered_tests, visible_name, res, i,
-                                            logfile, jsonlogfile)
+                        self.print_stats(numlen, tests, visible_name, res, i,
+                                logfile, jsonlogfile)
```

Comment:
__[flake8]__
* [E128] continuation line under-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->